### ISSUE TITLE
Testing assistive technology interoperability

### ISSUE THREAD

**Issue by dotherightthing** [w3c/aria-practices#1577]
Note? If this isn't the right place for this rant please let me know.

My perspective is as an accessibility advocate and a front-end developer contracting to design agencies, outside the US.

In [2.2 Browser and Assistive Technology Support](https://w3c.github.io/aria-practices/#browser_and_AT_support), there is a prominent disclaimer, which states that

> Testing assistive technology interoperability is essential before using code from this guide in production. Because the purpose of this guide is to illustrate appropriate use of ARIA 1.2 as defined in the ARIA specification, the design patterns, reference examples, and sample code intentionally do not describe and implement coding techniques for working around problems caused by gaps in support for ARIA 1.2 in browsers and assistive technologies. It is thus advisable to test implementations thoroughly with each browser and assistive technology combination that is relevant within a target audience.

Firstly, when I navigate to <https://w3c.github.io/aria-practices/>, I am taken to the ARIA 1.2 spec, not a different 'current' version. So I would assume from that this is an agreed standard, and has pretty good support. I wouldn't expect it to be so future focussed that it excludes users of older AT, since the whole purpose of web accessibility is to be more inclusive.

Secondly, I have major concerns about the way that responsibility for testing is being pushed onto implementors.

There are two ways that implementations can be tested:

1. Developer/implementor tests

* extremely problematic, especially if the developer is 'able' and not personally reliant on AT
* developer tests usability in some AT/browser combination (assuming they don't just run some automated checks)
* developer logs a bug that they think presents a barrier
* developer is not really sure if they're using the AT as a real user would, or as a 'power user' rather than a regular user
* line manager rejects or lowers priority of bug, noting developer's inexperience with AT, and/or
* line manager quotes the classic "but a ___ user would be used to (working around) that"
* developer is annoyed, but does not have the lifetime experience of exclusion and discrimination to give the line manager a properly convincing reality check

2. User tests

* really great for the insights but extremely problematic, because this very rarely happens, due to will/discrimination, cost, time, competing priorities
* WCAG emphasises technical testing over manual validation with real users, so developers have no mandate for pushing for user testing
* lots of different users to test, everyone has a different configuration of their PC/AT/browser which adds hundreds of variables beyond "each browser and assistive technology combination that is relevant within a target audience"
* many users don't have the latest devices/browsers/AT/training - so this excludes users who can't upgrade because they don't want to change their workflow / are reliant on state funding / work with carers who aren't technically literate

I really question the point of 'best practice' examples that ignore the realities of the real world out there. Yes, it's good to have a vision and a goal, no developer wants to be stuck supporting the AT equivalent of IE6. But IE6 was free, and came pre-installed on every computer. Nowadays we have a plethora of configurable devices/browsers/AT products.

At the end of the day, in lieu of sufficient knowledge or testing, developers are going to rely on (copy) the examples in the WAI-ARIA Authoring Practices.

If this creates a bad experience for some/many users, developers will not know.

Alternatively they might get some complaints via a support channel, read your disclaimer and realise that your examples are more idealistic than practical. At this point they will be forced onto Google, to find a better-practice solution from elsewhere. They will then choose from one of the many half-baked proof of concept 'solutions' proffered by well-meaning but time-poor accessibility advocates. 

Either way, the end result will be that the current situation won't change - developers will still think that they are nailing accessibility, while end users will still have crappy inconsistent experiences.

### TL;DR

1. I would like the WAI-ARIA Authoring Practices to be the go-to source for accessible widgets (if they aren't intended to be, then what are they?)
2. I would like the WAI-ARIA Authoring Practices to acknowledge real world requirements (legacy support, no user testing, proper setup of JavaScript-dependent stateful components using progressive enhancement)
3. I would like the ARIA technical experts to construct practical examples which factor in present, past and future users, the workarounds required for known browser/AT/configuration shortcomings, and a caniuse.com style matrix
